tools/autograd/gen_autograd_functions.py

```diff
@@ -84,142 +90,143 @@
-    }
-
-    for suffix in [".h", ".cpp"]:
-        f = file_basename + suffix
-        templated_output = CodeTemplate.from_file(os.path.join(template_path, f))
-        write(out, f, templated_output, top_env)
-
-
-def process_function(func):
-    env = {}
-    saved_variables = []
-    release_variables = []
-    saved_list_sizes = []
-    unpack = []
-    asserts = []
-
-    env['compute_index_ranges'] = []
-    for arg in func['args_with_derivatives']:
-        if arg['type'] == 'TensorList':
-            size = '{}_size_'.format(arg['name'])
-            saved_list_sizes.append('size_t {}_size_;'.format(arg['name']))
+    # only create an autograd function if we are actually going to calculate a derivative
+    infos = list(filter(lambda info: info.args_with_derivatives, differentiability_infos))
+    declarations = list(map(lambda f: process_function(f, FUNCTION_DECLARATION), infos))
+    definitions = list(map(lambda f: process_function(f, FUNCTION_DEFINITION), infos))
+    py_function_initializers = list(map(lambda f: process_function(f, PY_FUNCTION_DEFINITION), infos))
+
+    fm = FileManager(install_dir=out, template_dir=template_path, dry_run=False)
+    for suffix in ['.h', '.cpp']:
+        fname = file_basename + suffix
+        fm.write_with_template(fname, fname, lambda: {
+            'generated_comment': '@' + f'generated from {fm.template_dir}/' + fname,
+            'autograd_function_declarations': declarations,
+            'autograd_function_definitions': definitions,
+            'py_function_initializers': py_function_initializers,
+        })
+
+def process_function(info: DifferentiabilityInfo, template: CodeTemplate) -> str:
+    saved_variables: List[str] = []
+    release_variables: List[str] = []
+    saved_list_sizes: List[str] = []
+    unpack: List[str] = []
+    asserts: List[str] = []
+    compute_index_ranges: List[str] = []
+
+    for arg in info.args_with_derivatives:
+        if arg.type == 'TensorList':
+            size = f'{arg.name}_size_'
+            saved_list_sizes.append(f'size_t {arg.name}_size_;')
         else:
             size = '1'
-        env['compute_index_ranges'].append('auto {}_ix = gen.range({});'.format(arg['name'], size))
-
-    def save_arg(arg, is_output):
-        name = arg['name']
-
-        if arg['type'] == 'Tensor' or arg['type'] == 'c10::optional<Tensor>' or arg['type'] == 'c10::optional<Tensor>&' or \
-                (arg['type'] == 'Scalar' and is_output):
-            saved_variables.append('SavedVariable {}_;'.format(name))
-            release_variables.append('{}_.reset_data();'.format(name))
-            release_variables.append('{}_.reset_grad_function();'.format(name))
+        compute_index_ranges.append(f'auto {arg.name}_ix = gen.range({size});')
+
+    def save_var(var: Variable, is_output: bool) -> None:
+        name = var.name
+        if var.type == 'Tensor' or var.type == 'c10::optional<Tensor>' or var.type == 'c10::optional<Tensor>&' or \
```

Comment:
This is less refactor than I was hoping for, but this might actually be better since you're about to conflict with @albanD, the review has been pretty straightforward since things line up.

Reply:
The addition of forward AD might impose some more constraints. So I think we should wait on more refactor until it is there yes.